README
NAME
    TBX::Min - Read, write and edit TBX-Min files

VERSION
    version 0.06

SYNOPSIS
        use TBX::Min;
        my $min = TBX::Min->new('/path/to/file.tbx');
        my $entries = $min->entries;
        my $entry = TBX::Min::Entry->new({id => 'B001'});
        $min->add_entry($entry);

DESCRIPTION
    This module allows you to read, write and edit the contents of TBX-Min
    data.

    "use"ing this module also automatically "use"s TBX::Min::Entry,
    TBX::Min::LangGroup, and TBX::Min::TermGroup via Import::Into.
    LangGroups contain TermGroups, Entries contain LangGroups, and this
    class contains Entries. These correspond to the three levels of
    information found in TML. You can build up TBX::Min documents this way
    and then print them via "as_xml". You can also read an entire TBX-Min
    XML document for editing via "new_from_xml".

TBX-Min
    TBX-Min is a minimal, DCT-style dialect of TBX. It's purpose is to
    represent extremely simple termbases, such as spreadsheets, and to be as
    human eye-friendly as possible. TBX-Min did not evolve from any other
    XML dialect, and so does not have historical artifacts such as "martif".

    DCT stands for "Data Category as Tag Name". Whereas in most TBX dialects
    categories such as "partOfSpeech" are indicated through attributes, in
    TBX-Min the tag names represent categories. This makes for a very
    readable document. While TBX-Min documents do conform to TML
    (Terminological Markup Language) structure, DCT documents cannot be
    checked by the TBX-Checker <https://sourceforge.net/projects/tbxutil/>.

    If you need more complex or information-rich termbases, we suggest you
    use TBX-Basic or even TBX-Default. If you have a TBX-Min document and
    would like to upgrade it to TBX-Basic, see Convert::TBX::Min.
    Alternatively if you would like to change your TBX-Basic to TBX-Min, see
    Convert::TBX::Basic.

METHODS
  "new_from_xml"
    Creates a new instance of TBX::Min. The single argument should be either
    a string pointer containing the TBX-Min XML data or the name of the file
    containing this data is required.

  "new"
    Creates a new "TBX::Min" instance. Optionally you may pass in a hash
    reference which is used to initialize the object. The allowed hash
    fields are "id", "description", "date_created", "creator", "license",
    "directionality", "source_lang" and "target_lang", which correspond to
    methods of the same name, and "entries", which should be an array
    reference containing "TBX::Min::Entry" objects. This method croaks if
    "date_created" is not in ISO 8601 format.

  "id"
    Get or set the document id. This should be a unique string identifying
    this glossary.

  "description"
    Get or set the document description.

  "date_created"
    Get or set the the date that the document was created. This should be a
    string in ISO 8601 format. This method croaks if "date_created" is not
    in ISO 8601 format.

  "creator"
    Get or set the name of the document creator.

  "license"
    Get or set the document license string.

  "directionality"
    Get or set the document directionality string. This string represents
    the direction of translation this document is designed for.

  "source_lang"
    Get or set the code representing the document source language. This
    should be ISO 639 and 3166 (e.g. "en-US", "de", etc.).

  "target_lang"
    Get or set the code representing the document target language. This
    should be ISO 639 and 3166 (e.g. "en-US", "de", etc.).

  "entries"
    Returns an array ref containing the "TBX::Min::Entry" objects contained
    in the document.The array ref is the same one used to store the objects
    internally, so additions or removals from the array will be reflected in
    future calls to this method.

  "add_entry"
    Adds the input "TBX::Min::Entry" object to the list of language groups
    contained by this object.

  "as_xml"
    Returns a scalar reference containing an XML representation of this
    TBX-Min document. The data is a UTF-8 encoded string.

CAVEATS
    TBX::Min does not as of yet fully validate TBX-Min documents. It is
    possible to create non-validating XML via the as_xml method. This should
    be fixed in the future.

SEE ALSO
    The following related modules:

    TBX::Min::Entry
    TBX::Min::LangGroup
    TBX::Min::TermGroup
    Convert::TBX::Min
    Convert::TBX::Basic

    Schema for valiating TBX-Min files are available on GitHub
    <https://github.com/byutrg/TBX-Spec>.

AUTHOR
    Nathan Glenn <garfieldnate@gmail.com>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2013 by Alan Melby.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Revision history for TBX-Min

0.06      2014-02-03 13:01:04 America/Los_Angeles
    check input document for use of incorrect dialect
    import TBX::Min::Entry, etc. automatically
    updated documentation
0.05      2014-01-30 14:55:25 America/Los_Angeles
    Fixed warning on 5.10.X
    add_concept changed to add_entry as should have been done in 0.04
    as_xml returns pretty-printed XML
    as_xml return scalar ref instead of just scalar
0.04      2014-01-28 19:27:39 America/Los_Angeles
    "conceptEntry" element changed to "entry"
0.03      2013-11-19 16:50:25 America/Los_Angeles
    Uninitialized arrays are now returned as [] instead of undef
0.02      2013-11-12 14:07:47 America/Los_Angeles
    Several changes were made to the TBX-Min dialect to ensure compatibility
    with UTX:
        renamed "origin" to "creator"
        renamed "title" to "id"
        added "description" field to header
        added "dateCreated" field to header
        removed "doc_lang"

0.01      2013-11-08 21:18:18 America/Los_Angeles
	initial release


Build.PL

use strict;
use warnings;

use Module::Build 0.3601;


my %module_build_args = (
  "build_requires" => {
    "Module::Build" => "0.3601"
  },
  "configure_requires" => {
    "Module::Build" => "0.3601"
  },
  "dist_abstract" => "Read, write and edit TBX-Min files\r",
  "dist_author" => [
    "Nathan Glenn <garfieldnate\@gmail.com>"
  ],
  "dist_name" => "TBX-Min",
  "dist_version" => "0.06",
  "license" => "perl",
  "module_name" => "TBX::Min",
  "recommends" => {},
  "recursive_test_files" => 1,
  "requires" => {
    "Carp" => 0,
    "Class::Tiny" => 0,
    "DateTime::Format::ISO8601" => 0,
    "Import::Into" => 0,
    "Path::Tiny" => 0,
    "Try::Tiny" => 0,
    "XML::Twig" => 0,
    "autodie" => 0,
    "perl" => "5.006",
    "strict" => 0,
    "subs" => 0,
    "warnings" => 0
  },
  "script_files" => [],
  "test_requires" => {
    "File::Find" => 0,
    "File::Temp" => 0,
    "FindBin" => 0,
    "Test::Deep" => 0,
    "Test::Exception" => 0,
    "Test::More" => "0.88",
    "Test::NoWarnings" => 0,
    "Test::XML" => 0
  }
);


unless ( eval { Module::Build->VERSION(0.4004) } ) {
  my $tr = delete $module_build_args{test_requires};
  my $br = $module_build_args{build_requires};
  for my $mod ( keys %$tr ) {
    if ( exists $br->{$mod} ) {
      $br->{$mod} = $tr->{$mod} if $tr->{$mod} > $br->{$mod};
    }
    else {
      $br->{$mod} = $tr->{$mod};
    }
  }
}

my $build = Module::Build->new(%module_build_args);

$build->create_build_script;


Makefile.PL

use strict;
use warnings;

use 5.006;

use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "Read, write and edit TBX-Min files\r",
  "AUTHOR" => "Nathan Glenn <garfieldnate\@gmail.com>",
  "BUILD_REQUIRES" => {
    "Module::Build" => "0.3601"
  },
  "CONFIGURE_REQUIRES" => {
    "Module::Build" => "0.3601"
  },
  "DISTNAME" => "TBX-Min",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "NAME" => "TBX::Min",
  "PREREQ_PM" => {
    "Carp" => 0,
    "Class::Tiny" => 0,
    "DateTime::Format::ISO8601" => 0,
    "Import::Into" => 0,
    "Path::Tiny" => 0,
    "Try::Tiny" => 0,
    "XML::Twig" => 0,
    "autodie" => 0,
    "strict" => 0,
    "subs" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "File::Find" => 0,
    "File::Temp" => 0,
    "FindBin" => 0,
    "Test::Deep" => 0,
    "Test::Exception" => 0,
    "Test::More" => "0.88",
    "Test::NoWarnings" => 0,
    "Test::XML" => 0
  },
  "VERSION" => "0.06",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  my $tr = delete $WriteMakefileArgs{TEST_REQUIRES};
  my $br = $WriteMakefileArgs{BUILD_REQUIRES};
  for my $mod ( keys %$tr ) {
    if ( exists $br->{$mod} ) {
      $br->{$mod} = $tr->{$mod} if $tr->{$mod} > $br->{$mod};
    }
    else {
      $br->{$mod} = $tr->{$mod};
    }
  }
}

unless ( eval { ExtUtils::MakeMaker->VERSION(6.56) } ) {
  my $br = delete $WriteMakefileArgs{BUILD_REQUIRES};
  my $pp = $WriteMakefileArgs{PREREQ_PM};
  for my $mod ( keys %$br ) {
    if ( exists $pp->{$mod} ) {
      $pp->{$mod} = $br->{$mod} if $br->{$mod} > $pp->{$mod};
    }
    else {
      $pp->{$mod} = $br->{$mod};
    }
  }
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);





lib/TBX/Min/Entry.pm
#
# This file is part of TBX-Min
#
# This software is copyright (c) 2013 by Alan Melby.
#
# This is free software; you can redistribute it and/or modify it under
# the same terms as the Perl 5 programming language system itself.
#
package TBX::Min::Entry;
use strict;
use warnings;
use Carp;
our $VERSION = '0.06'; # VERSION

# ABSTRACT: Store information from one TBX-Min C<entry> element

sub new {
    my ($class, $args) = @_;
    my $self;
    if((ref $args) eq 'HASH'){
        $self = $args;
    }else{
        $self = {};
    }
    $self->{lang_groups} ||= [];
    return bless $self, $class;
}

sub id {
    my ($self, $id) = @_;
    if($id) {
        return $self->{id} = $id;
    }
    return $self->{id};
}

sub subject_field {
    my ($self, $subject_field) = @_;
    if($subject_field) {
        return $self->{subject_field} = $subject_field;
    }
    return $self->{subject_field};
}

sub lang_groups { ## no critic(RequireArgUnpacking)
    my ($self) = @_;
    if (@_ > 1){
        croak 'extra argument found (lang_groups is a getter only)';
    }
    return $self->{lang_groups};
}

sub add_lang_group {
    my ($self, $lang_grp) = @_;
    if( !$lang_grp || !$lang_grp->isa('TBX::Min::LangGroup') ){
        croak 'argument to add_lang_group should be a TBx::Min::LangGroup';
    }
    push @{$self->{lang_groups}}, $lang_grp;
    return;
}


1;

__END__

=pod

=head1 NAME

TBX::Min::Entry - Store information from one TBX-Min C<entry> element

=head1 VERSION

version 0.06

=head1 SYNOPSIS

    use TBX::Min::Entry;
    use TBX::Min::LangGroup;
    my $concept = TBX::Min::Entry->new(
        {id => 'B001'});
    print $concept->id(); # 'B001'
    my $lang_grp = TBX::Min::LangGroup->new({code => 'en'});
    $concept->add_lang_group($lang_grp);
    my $lang_grps = $concept->lang_groups;
    print $#$lang_grps; # '1'

=head1 DESCRIPTION

This class represents a single concept entry contained in a TBX-Min file. A
concept entry contains several language groups, each containing terms that
represent this concept in a given languages.

=head1 METHODS

=head2 C<new>

Creates a new C<TBX::Min::Entry> instance. Optionally you may pass in
a hash reference which is used to initialize the object. The allowed hash
fields are C<id>, C<subject_field> and C<lang_groups>, where C<id> and
C<subject_field> correspond to methods of the same name, and C<langGroups> is
an array reference containing C<TBX::Min::LangGroup> objects.

=head2 C<id>

Get or set the concept entry id value.

=head2 C<subject_field>

Get or set the concept subject field string.

=head2 C<lang_groups>

Returns an array ref containing all of the C<TBX::Min::LangGroup> objects
in this concept entry. The array ref is the same one used to store the objects
internally, so additions or removals from the array will be reflected in future
calls to this method.

=head2 C<add_lang_group>

Adds the input C<TBX::Min::LangGroup> object to the list of language groups
contained by this object.

=head1 SEE ALSO

L<TBX::Min>

=head1 AUTHOR

Nathan Glenn <garfieldnate@gmail.com>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2013 by Alan Melby.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


lib/TBX/Min/LangGroup.pm
#
# This file is part of TBX-Min
#
# This software is copyright (c) 2013 by Alan Melby.
#
# This is free software; you can redistribute it and/or modify it under
# the same terms as the Perl 5 programming language system itself.
#
package TBX::Min::LangGroup;
use strict;
use warnings;
use Carp;
our $VERSION = '0.06'; # VERSION

# ABSTRACT: Store information from one TBX-Min C<langGroup> element

sub new {
    my ($class, $args) = @_;
    my $self;
    if((ref $args) eq 'HASH'){
        $self = $args;
    }else{
        $self = {};
    }
    $self->{term_groups} ||= [];
    return bless $self, $class;
}

sub code {
    my ($self, $code) = @_;
    if($code) {
        return $self->{code} = $code;
    }
    return $self->{code};
}

sub term_groups { ## no critic(RequireArgUnpacking)
    my ($self) = @_;
    if (@_ > 1){
        croak 'extra argument found (term_groups is a getter only)';
    }
    return $self->{term_groups};
}

sub add_term_group {
    my ($self, $term_grp) = @_;
    if( !$term_grp || !$term_grp->isa('TBX::Min::TermGroup') ){
        croak 'argument to add_term_group should be a TBx::Min::TermGroup';
    }
    push @{$self->{term_groups}}, $term_grp;
    return;
}


1;

__END__

=pod

=head1 NAME

TBX::Min::LangGroup - Store information from one TBX-Min C<langGroup> element

=head1 VERSION

version 0.06

=head1 SYNOPSIS

    use TBX::Min::LangGroup;
    use TBX::Min::TermGroup;
    my $lang_grp = TBX::Min::LangGroup->new(
        {code => 'en'});
    print $lang_grp->lang(); # 'en'
    my $term_grp = TBX::Min::TermGroup->new({term => 'perl'});
    $lang_grp->add_term_group($term_grp);
    my $term_grps = $lang_grp->term_groups;
    print $#$term_grps; # '1'

=head1 DESCRIPTION

This class represents a single language group contained in a TBX-Min file.
A language group is contained by a concept entry, and contains several term
groups each representing a given concept for the same language.

=head1 METHODS

=head2 C<new>

Creates a new C<TBX::Min::LangGroup> instance. Optionally you may pass in
a hash reference which is used to initialize the object. The allowed hash
fields are C<code> and C<term_groups>, where C<code> corresponds to the
method of the same name, and C<term_groups> is an array reference containing
C<TBX::Min::LangGroup> objects.

=head2 C<code>

Get or set the language group language code (should be ISO 639 and 3166,
e.g. C<en-US>, C<de>, etc.).

=head2 C<term_groups>

Returns an array ref containing all of the C<TBX::Min::TermGroup> objects
in this concept entry. The array ref is the same one used to store the objects
internally, so additions or removals from the array will be reflected in future
calls to this method.

=head2 C<add_term_group>

Adds the input C<TBX::Min::TermGroup> object to the list of language groups
contained by this object.

=head1 SEE ALSO

L<TBX::Min>

=head1 AUTHOR

Nathan Glenn <garfieldnate@gmail.com>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2013 by Alan Melby.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


lib/TBX/Min/TermGroup.pm
#
# This file is part of TBX-Min
#
# This software is copyright (c) 2013 by Alan Melby.
#
# This is free software; you can redistribute it and/or modify it under
# the same terms as the Perl 5 programming language system itself.
#
package TBX::Min::TermGroup;
use strict;
use warnings;
use subs qw(part_of_speech status);
use Class::Tiny qw(
    term
    part_of_speech
    note
    customer
    status
);
use Carp;

our $VERSION = '0.06'; # VERSION

# ABSTRACT: Store information from one TBX-Min C<termGroup> element


sub part_of_speech {
    my ($self, $pos) = @_;
    if(defined $pos){
        _validate_pos($pos);
        $self->{part_of_speech} = $pos;
    }
    return $self->{part_of_speech};
}


sub status {
    my ($self, $status) = @_;
    if(defined $status){
        _validate_status($status);
        $self->{status} = $status;
    }
    return $self->{status};
}

# above Pod::Coverage makes this not "naked" via Pod::Coverage::TrustPod
sub BUILD {
    my ($self, $args) = @_;
    if($args->{part_of_speech}){
        _validate_pos($args->{part_of_speech});
    }
    if($args->{status}){
        _validate_status($args->{status});
    }
    return;
}

my @allowed_pos = qw(noun properNoun verb adjective adverb other);
sub _validate_pos {
    my ($pos) = @_;
    if(!grep{$pos eq $_} @allowed_pos){
        croak "Illegal part of speech '$pos'";
    }
    return;
}

my @allowed_status = qw(admitted preferred notRecommended obsolete);
sub _validate_status {
    my ($pos) = @_;
    if(!grep{$pos eq $_} @allowed_status){
        croak "Illegal status '$pos'";
    }
    return;
}

1;

__END__

=pod

=head1 NAME

TBX::Min::TermGroup - Store information from one TBX-Min C<termGroup> element

=head1 VERSION

version 0.06

=head1 SYNOPSIS

    use TBX::Min::TermGroup;
    my $term_grp = TBX::Min::TermGroup->new(
        {term => 'bat signal', status => "preferred"});
    $term_grp->part_of_speech('noun');
    $term_grp->customer('GCPD');
    print $term_grp->term; # 'bat signal'

=head1 DESCRIPTION

This class represents a single term group contained in a TBX-Min file. A term
group contains a single term and information pertaining to it, such as part of
speech, a note, or the associated customer.

=head1 METHODS

=head2 C<new>

Creates a new C<TBX::Min::TermGroup> instance. Optionally you may pass in a hash
reference which is used to initialized the object. The fields of the hash
correspond to the names of the accessor methods listed below.

=head2 C<term>

Get or set the term text associated with this term group.

=head2 C<part_of_speech>

Get or set the part of speech associated with this term group.

=head2 C<note>

Get or set a note associated with this term group.

=head2 C<customer>

Get or set a customer associated with this term group.

=head2 C<status>

Get or set a status  associated with this term group.

=head1 SEE ALSO

L<TBX::Min>

=for Pod::Coverage BUILD

=head1 AUTHOR

Nathan Glenn <garfieldnate@gmail.com>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2013 by Alan Melby.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


lib/TBX/Min.pm
#
# This file is part of TBX-Min
#
# This software is copyright (c) 2013 by Alan Melby.
#
# This is free software; you can redistribute it and/or modify it under
# the same terms as the Perl 5 programming language system itself.
#
package TBX::Min;
use strict;
use warnings;
our $VERSION = '0.06'; # VERSION
# ABSTRACT: Read, write and edit TBX-Min files
use XML::Twig;
use autodie;
use Path::Tiny;
use Carp;
use TBX::Min::Entry;
use TBX::Min::LangGroup;
use TBX::Min::TermGroup;
use Import::Into;
use DateTime::Format::ISO8601;
use Try::Tiny;

# Use Import::Into to export subclasses into caller
sub import {
    my $target = caller;
    TBX::Min::Entry->import::into($target);
    TBX::Min::LangGroup->import::into($target);
    TBX::Min::TermGroup->import::into($target);
    return;
}

sub new_from_xml {
    my ($class, $data) = @_;

    if(!$data){
        croak 'missing required data argument';
    }

    my $fh = _get_handle($data);

    # build a twig out of the input document
    my $twig = XML::Twig->new(
        output_encoding => 'UTF-8',
        # do_not_chain_handlers => 1, #can be important when things get complicated
        keep_spaces     => 0,

        # these store new entries, langGroups and termGroups
        start_tag_handlers => {
            entry => \&_conceptStart,
            langGroup => \&_langStart,
            termGroup => \&_termGrpStart,
        },

        TwigHandlers    => {
            TBX => \&_check_dialect,
            # header attributes become attributes of the TBX::Min object
            id => \&_headerAtt,
            description => \&_headerAtt,
            dateCreated => \&_date_created,
            creator => \&_headerAtt,
            license => \&_headerAtt,
            directionality => \&_directionality,
            languages => \&_languages,

            # becomes part of the current TBX::Min::Entry object
            subjectField => sub {
                shift->{tbx_min_entries}->[-1]->subject_field($_->text)},

            # these become attributes of the current TBX::Min::TermGroup object
            term => sub {shift->{tbx_min_current_term_grp}->term($_->text)},
            partOfSpeech => sub {
                shift->{tbx_min_current_term_grp}->part_of_speech($_->text)},
            note => sub {shift->{tbx_min_current_term_grp}->note($_->text)},
            customer => sub {
                shift->{tbx_min_current_term_grp}->customer($_->text)},
            termStatus => sub {
                shift->{tbx_min_current_term_grp}->status($_->text)},
        }
    );

    # use handlers to process individual tags, then grab the result
    $twig->parse($fh);
    my $self = $twig->{tbx_min_att};
    $self->{entries} = $twig->{tbx_min_entries} || [];
    bless $self, $class;
    return $self;
}

sub _get_handle {
    my ($data) = @_;
    my $fh;
    if((ref $data) eq 'SCALAR'){
        open $fh, '<', $data; ## no critic(RequireBriefOpen)
    }else{
        $fh = path($data)->filehandle('<');
    }
    return $fh;
}

sub new {
    my ($class, $args) = @_;
    my $self;
    if((ref $args) eq 'HASH'){
        #don't store a plain string for datetime
        if(my $dt_string = $args->{date_created}){
            $args->{date_created} = _parse_datetime($dt_string);
        }
        if(exists $args->{directionality}){
            _validate_dir($args->{directionality});
        }
        $self = $args;
    }else{
        $self = {};
    }
    $self->{entries} ||= [];
    return bless $self, $class;
}

sub id {
    my ($self, $id) = @_;
    if($id) {
        return $self->{id} = $id;
    }
    return $self->{id};
}

sub description {
    my ($self, $description) = @_;
    if($description) {
        return $self->{description} = $description;
    }
    return $self->{description};
}

sub date_created {
    my ($self, $date_created) = @_;
    if($date_created) {
        return $self->{date_created} =
            _parse_datetime($date_created);
    }
    if(my $dt = $self->{date_created}){
        return $dt->iso8601;
    }
    return;
}

sub _parse_datetime {
    my ($dt_string) = @_;
    my $dt;
    try{
        $dt = DateTime::Format::ISO8601->parse_datetime($dt_string);
    }catch{
        croak 'date is not in ISO8601 format';
    };
    return $dt;
}

sub creator {
    my ($self, $creator) = @_;
    if($creator) {
        return $self->{creator} = $creator;
    }
    return $self->{creator};
}

sub license {
    my ($self, $license) = @_;
    if($license) {
        return $self->{license} = $license;
    }
    return $self->{license};
}

sub directionality {
    my ($self, $directionality) = @_;
    if(defined $directionality) {
        _validate_dir($directionality);
        return $self->{directionality} = $directionality;
    }
    return $self->{directionality};
}

sub _validate_dir {
    my ($dir) = @_;
    if($dir ne 'bidirectional' and $dir ne 'monodirectional'){
        croak "Illegal directionality '$dir'";
    }
    return;
}


sub source_lang {
    my ($self, $source_lang) = @_;
    if($source_lang) {
        return $self->{source_lang} = $source_lang;
    }
    return $self->{source_lang};
}

sub target_lang {
    my ($self, $target_lang) = @_;
    if($target_lang) {
        return $self->{target_lang} = $target_lang;
    }
    return $self->{target_lang};
}

sub entries { ## no critic(RequireArgUnpacking)
    my ($self) = @_;
    if (@_ > 1){
        croak 'extra argument found (entries is a getter only)';
    }
    return $self->{entries};
}

sub add_entry {
    my ($self, $entry) = @_;
    if( !$entry || !$entry->isa('TBX::Min::Entry') ){
        croak 'argument to add_entry should be a TBx::Min::Entry';
    }
    push @{$self->{entries}}, $entry;
    return;
}

sub as_xml {
    my ($self) = @_;

    # construct the whole document using XML::Twig::El's
    my $root = XML::Twig::Elt->new(TBX => {dialect => 'TBX-Min'});
    my $header = XML::Twig::Elt->new('header')->paste($root);

    # each of these header elements is a simple element with text
    for my $header_att (
            qw(id creator license directionality description)){
        next unless $self->{$header_att};
        XML::Twig::Elt->new($header_att,
            $self->{$header_att})->paste(last_child => $header);
    }
    if($self->source_lang || $self->target_lang){
        my @atts;
        push @atts, (source => $self->source_lang) if $self->source_lang;
        push @atts, (target => $self->target_lang) if $self->target_lang;
        XML::Twig::Elt->new(languages => {@atts})->paste(
            last_child => $header)
    }
    if(my $dt = $self->{date_created}){
        XML::Twig::Elt->new(dateCreated => $dt->iso8601)->paste(
            last_child => $header);
    }

    my $body = XML::Twig::Elt->new('body')->paste(last_child => $root);
    for my $entry (@{$self->entries}){
        my $entry_el = XML::Twig::Elt->new(
            entry => {$entry->id ? (id => $entry->id) : ()})->
            paste(last_child => $body);
        if(my $sf = $entry->subject_field){
            XML::Twig::Elt->new(subjectField => $sf)->paste(
                last_child => $entry_el);
        }
        for my $langGrp (@{$entry->lang_groups}){
            my $lang_el = XML::Twig::Elt->new(langGroup =>
                {$langGrp->code ? ('xml:lang' => $langGrp->code) : ()}
            )->paste(last_child => $entry_el);
            for my $termGrp (@{$langGrp->term_groups}){
                my $term_el = XML::Twig::Elt->new('termGroup')->paste(
                    last_child => $lang_el);
                if (my $term = $termGrp->term){
                    XML::Twig::Elt->new(term => $term)->paste(
                        last_child => $term_el);
                }

                if (my $customer = $termGrp->customer){
                    XML::Twig::Elt->new(customer => $customer)->paste(
                        last_child => $term_el);
                }

                if (my $note = $termGrp->note){
                    XML::Twig::Elt->new(note => $note)->paste(
                        last_child => $term_el);
                }

                if (my $status = $termGrp->status){
                    XML::Twig::Elt->new(termStatus => $status )->paste(
                        last_child => $term_el);
                }

                if (my $pos = $termGrp->part_of_speech){
                    XML::Twig::Elt->new(partOfSpeech => $pos)->paste(
                        last_child => $term_el);
                }

            } # end termGroup
        } # end langGroup
    } # end entry

    # return pretty-printed string
    XML::Twig->set_pretty_print('indented');
    return \$root->sprint;
}

######################
### XML TWIG HANDLERS
######################

# croak if the user happened to use the wrong dialect of TBX
sub _check_dialect {
    my (undef, $node) = @_;
    my $type = $node->att('dialect') || 'unknown';
    my $expected = 'TBX-Min';
    if($type ne $expected){
        croak "Input TBX is $type (should be '$expected')";
    }
    return 1;
}

# most of the twig handlers store state on the XML::Twig object.
# A bit kludgy, but it works.

sub _headerAtt {
    my ($twig, $node) = @_;
    $twig->{tbx_min_att}->{_decamel($node->name)} = $node->text;
    return 1;
}

sub _directionality {
    my ($twig, $node) = @_;
    _validate_dir($node->text);
    $twig->{tbx_min_att}->{directionality} = $node->text;
    return 1;
}

sub _date_created {
    my ($twig, $node) = @_;
    $twig->{tbx_min_att}->{date_created} =
        _parse_datetime($node->text);
    return;
}

# turn camelCase into camel_case
sub _decamel {
    my ($camel) = @_;
    $camel =~ s/([A-Z])/_\l$1/g;
    return $camel;
}

sub _languages{
    my ($twig, $node) = @_;
    if(my $source = $node->att('source')){
        ${ $twig->{'tbx_min_att'} }{'source_lang'} = $source;
    }
    if(my $target = $node->att('target')){
        ${ $twig->{'tbx_min_att'} }{'target_lang'} = $target;
    }
    return 1;
}

# add a new concept entry to the list of those found in this file
sub _conceptStart {
    my ($twig, $node) = @_;
    my $concept = TBX::Min::Entry->new();
    if($node->att('id')){
        $concept->id($node->att('id'));
    }else{
        carp 'found entry missing id attribute';
    }
    push @{ $twig->{tbx_min_entries} }, $concept;
    return 1;
}

#just set the subject_field of the current concept
sub _subjectField {
    my ($twig, $node) = @_;
    $twig->{tbx_min_entries}->[-1]->
        subject_field($node->text);
    return 1;
}

# Create a new LangGroup, add it to the current concept,
# and set it as the current LangGroup.
sub _langStart {
    my ($twig, $node) = @_;
    my $lang = TBX::Min::LangGroup->new();
    if($node->att('xml:lang')){
        $lang->code($node->att('xml:lang'));
    }else{
        carp 'found langGroup missing xml:lang attribute';
    }

    $twig->{tbx_min_entries}->[-1]->add_lang_group($lang);
    $twig->{tbx_min_current_lang_grp} = $lang;
    return 1;
}

# Create a new termGroup, add it to the current langGroup,
# and set it as the current termGroup.
sub _termGrpStart {
    my ($twig) = @_;
    my $term = TBX::Min::TermGroup->new();
    $twig->{tbx_min_current_lang_grp}->add_term_group($term);
    $twig->{tbx_min_current_term_grp} = $term;
    return 1;
}

1;

__END__

=pod

=head1 NAME

TBX::Min - Read, write and edit TBX-Min files

=head1 VERSION

version 0.06

=head1 SYNOPSIS

    use TBX::Min;
    my $min = TBX::Min->new('/path/to/file.tbx');
    my $entries = $min->entries;
    my $entry = TBX::Min::Entry->new({id => 'B001'});
    $min->add_entry($entry);

=head1 DESCRIPTION

This module allows you to read, write and edit the contents of TBX-Min
data.

C<use>ing this module also automatically C<use>s L<TBX::Min::Entry>,
L<TBX::Min::LangGroup>, and L<TBX::Min::TermGroup> via
L<Import::Into>. LangGroups contain TermGroups, Entries contain
LangGroups, and this class contains Entries. These correspond to the
three levels of information found in TML. You can build up TBX::Min
documents this way and then print them via L</as_xml>. You can also
read an entire TBX-Min XML document for editing via L</new_from_xml>.

=head1 TBX-Min

TBX-Min is a minimal, DCT-style dialect of TBX. It's purpose is to
represent extremely simple termbases, such as spreadsheets, and to
be as human eye-friendly as possible. TBX-Min did not evolve from
any other XML dialect, and so does not have historical artifacts
such as "martif".

DCT stands for "Data Category as Tag Name". Whereas in most TBX
dialects categories such as C<partOfSpeech> are indicated through
attributes, in TBX-Min the tag names represent categories. This
makes for a very readable document. While TBX-Min documents do
conform to TML (Terminological Markup Language) structure, DCT
documents cannot be checked by the
L<TBX-Checker|https://sourceforge.net/projects/tbxutil/>.

If you need more complex or information-rich termbases, we suggest
you use TBX-Basic or even TBX-Default. If you have a TBX-Min document
and would like to upgrade it to TBX-Basic, see L<Convert::TBX::Min>.
Alternatively if you would like to change your TBX-Basic to TBX-Min,
see L<Convert::TBX::Basic>.

=head1 METHODS

=head2 C<new_from_xml>

Creates a new instance of TBX::Min. The single argument should be either a
string pointer containing the TBX-Min XML data or the name of the file
containing this data is required.

=head2 C<new>

Creates a new C<TBX::Min> instance. Optionally you may pass in
a hash reference which is used to initialize the object. The allowed hash
fields are C<id>, C<description>, C<date_created>, C<creator>, C<license>,
C<directionality>, C<source_lang> and C<target_lang>, which correspond to
methods of the same name, and C<entries>, which should be an array reference
containing C<TBX::Min::Entry> objects. This method croaks if
C<date_created> is not in ISO 8601 format.

=head2 C<id>

Get or set the document id. This should be a unique string
identifying this glossary.

=head2 C<description>

Get or set the document description.

=head2 C<date_created>

Get or set the the date that the document was created. This should be a
string in ISO 8601 format. This method croaks if C<date_created> is not
in ISO 8601 format.

=head2 C<creator>

Get or set the name of the document creator.

=head2 C<license>

Get or set the document license string.

=head2 C<directionality>

Get or set the document directionality string. This string represents
the direction of translation this document is designed for.

=head2 C<source_lang>

Get or set the code representing the document source language. This should
be ISO 639 and 3166 (e.g. C<en-US>, C<de>, etc.).

=head2 C<target_lang>

Get or set the code representing the document target language. This should
be ISO 639 and 3166 (e.g. C<en-US>, C<de>, etc.).

=head2 C<entries>

Returns an array ref containing the C<TBX::Min::Entry> objects contained
in the document.The array ref is the same one used to store the objects
internally, so additions or removals from the array will be reflected in future
calls to this method.

=head2 C<add_entry>

Adds the input C<TBX::Min::Entry> object to the list of language groups
contained by this object.

=head2 C<as_xml>

Returns a scalar reference containing an XML representation of this
TBX-Min document. The data is a UTF-8 encoded string.

=head1 CAVEATS

TBX::Min does not as of yet fully validate TBX-Min documents. It is
possible to create non-validating XML via the L<as_xml> method.
This should be fixed in the future.

=head1 SEE ALSO

The following related modules:

=over

=item L<TBX::Min::Entry>

=item L<TBX::Min::LangGroup>

=item L<TBX::Min::TermGroup>

=item L<Convert::TBX::Min>

=item L<Convert::TBX::Basic>

=back

Schema for valiating TBX-Min files are available on
L<GitHub|https://github.com/byutrg/TBX-Spec>.

=head1 AUTHOR

Nathan Glenn <garfieldnate@gmail.com>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2013 by Alan Melby.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut
